libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/cx_built_in_test.lua

```diff
@@ -1,242 +1,143 @@
-local function protected_wrapper()
-    local success, err = pcall(update)
-    if not success then
-        gcs_msg(MAV_SEVERITY_ERROR, "Internal Error: " .. err)
-        -- when we fault we run the update function again after 1s, slowing it
-        -- down a bit so we don't flood the console with errors
-        return protected_wrapper, 1000
-    end
-    return protected_wrapper, 200
-end
-
-local function script_exit()
-    -- pre arm failure SCRIPT_NAME not Running
-    arming:set_aux_auth_failed(auth_id, SCRIPT_NAME .. " Not Running")
-    gcs_msg(MAV_SEVERITY_CRITICAL, "LUA SCRIPT EXIT   ... Need Reboot to Reinitialize")
-end
-
-
--- ******************* Main *******************
-if arming_check_init() then
-    return protected_wrapper, 10000
-end
-
-script_exit()
+local cx_msg = require("msg")
+local cx_esc = require("bit_esc")
+
+-- Add subsystems that require Built-in-test (implemented in subsystems)
+-- Each subsystem should have the following functions:
+-- 1. init() - initialize the subsystem
+-- 2. update() - update the subsystem
+-- 3. check_for_errors() - returns pre-arm checks/errors in the subsystem
+--                       - built in test errors are managed by each subsystem
+local subsystems = {
+    cx_esc,
+}
+
+-- auth id for prearm check
+local prearm_msg = nil
+local auth_id = arming:get_aux_auth_id()
+assert(auth_id, SCRIPT_NAME .. ": could not get prearm check auth id")
+
+-- ******************* Functions *******************
+-- get time in seconds since boot
+local function get_time()
+    return millis():tofloat() * 0.001
+end
+
+local function set_prearm_error(txt)
+    if (not prearm_msg) or (prearm_msg ~= txt) then
+        prearm_msg = txt
+        arming:set_aux_auth_failed(auth_id, txt)
+    end
+end
+
+local function clear_prearm_error()
+    prearm_msg = nil
+    arming:set_aux_auth_passed(auth_id)
+end
+
+-- initialize function
+local function init()
+    -- initialize all subsystems that are part of constructor
+    for _, subsystem in pairs(subsystems) do
+        subsystem:init()
+    end
+
+    cx_msg:send(cx_msg.MAV_SEVERITY.INFO, "LUA script initialized")
+    return true
+end
+
+-- Pre-arm status check before arming
+local last_prearm_msg_s = 0 -- timestamp of last message sent
+local prearm_messages = {} -- Set of all unique messages seen since we last sent
+local function check_prearm_status()
+    -- Track errors in subsystems
+    local subsystems_with_errors = {}
+    local msg = ""
+    for _, subsystem in pairs(subsystems) do
+        local errors = subsystem:check_for_errors()
+        if #errors > 0 then
+            table.insert(subsystems_with_errors, subsystem.name)
+            for _, error in pairs(errors) do
+                prearm_messages[error] = true
+            end
+            if #errors == 1 then
+                msg = errors[1]
+            else
+                msg = errors .. " " .. subsystem.name " errors. Check messages."
```

Comment:
```suggestion
                msg = errors .. " " .. subsystem.name .. " errors. Check messages."
```

Reply:
@Pradeep-Carbonix @robertlong13

Reply:
Crap, good catch.

Reply:
And I didn't catch that in testing because currently all my subsystems only ever send one error.